### PR TITLE
IBP 3867 Create Lots from Crosses and selections Tab

### PR DESCRIPTION
**Pull request by abatac**

### CODE REVIEW

File: src/main/java/org/generationcp/middleware/dao/GermplasmStudySourceDAO.java [hunk 1]
```diff
@@ -52,7 +56,16 @@
 		+ "LEFT JOIN ims_lot lot ON lot.eid = gss.gid \n"
 		+ "WHERE gss.project_id = :studyId ";
 
-	public List<GermplasmStudySourceDto> getGermplasmStudySourceList(final GermplasmStudySourceSearchRequest germplasmStudySourceSearchRequest) {
+	public Map<Integer, GermplasmStudySource> getGermplasmStudySourcesMap(final Set<Integer> gids) {
+		final Criteria criteria = this.getSession().createCriteria(this.getPersistentClass());
+		criteria.createAlias("germplasm", "germplasm");
```
Comment (darla-leafnode) [inline src/main/java/org/generationcp/middleware/dao/GermplasmStudySourceDAO.java:61]: Do we have a unique key on `germplasm_study_source.gid` field?

File: src/main/java/org/generationcp/middleware/domain/inventory/manager/LotDepositRequestDto.java [hunk 2]
```diff
@@ -15,6 +15,8 @@
 
 	private String notes;
 
+	private Integer sourceStudy;
```
Comment (darla-leafnode) [inline src/main/java/org/generationcp/middleware/domain/inventory/manager/LotDepositRequestDto.java:18]: Should we rename to `sourceStudyId`?

File: src/main/java/org/generationcp/middleware/dao/GermplasmStudySourceDAO.java [hunk 3]
```diff
@@ -52,7 +56,16 @@
...
+		criteria.add(Restrictions.in("germplasm.gid", gids));
+		final List<GermplasmStudySource> result = criteria.list();
+		return result.stream().collect(Collectors.toMap(a -> a.getGermplasm().getGid(), Function.identity()));
```
Comment (darla-leafnode) [inline src/main/java/org/generationcp/middleware/dao/GermplasmStudySourceDAO.java:64]: For me, it's better to build the map by GID in the service instead in DAO class so you can reuse for other purposes.
As such, you can rename to public `LIst<GermplasmStudySource> getByGids(final Set<Integer gids>)`

File: src/main/java/org/generationcp/middleware/service/impl/inventory/TransactionServiceImpl.java [hunk 4]
```diff
@@ -200,11 +208,28 @@ public void depositLots(final Integer userId, final Set<Integer> lotIds, final L
...
+				// Create experiment transaction records when lot and deposit are created in the context of study.
+				this.createExperimentTransaction(lotDto.getGid(), germplasmStudySourceMap, transaction);
+			}
+
 		}
 
 	}
 
+	private void createExperimentTransaction(final Integer gid, final Map<Integer, GermplasmStudySource> germplasmStudySourceMap,
+		final Transaction transaction) {
```
Comment (darla-leafnode) [inline src/main/java/org/generationcp/middleware/service/impl/inventory/TransactionServiceImpl.java:223]: Probably offer `ExperimentTransactionType` as a parameter so as not to hardcode to `HARVESTING` type

File: src/main/resources/liquibase/workbench_changelog/15_4_0.xml [hunk 5]
```diff
@@ -22,4 +22,24 @@
...
+            <sqlCheck expectedResult="0">
+                SELECT count(*) FROM permission where name = 'MS_CREATE_LOTS'
+			</sqlCheck>
+        </preConditions>
+        <comment>Create granular permissions for Manage studies: Create Lots</comment>
+		<sql dbms="mysql" splitStatements="true">
+			set @manage_studies_id = (select permission_id from permission where name = 'MANAGE_STUDIES');
+			set @sidebar_category_link_id = (select sidebar_category_link_id from workbench_sidebar_category_link where tool_name = 'study_manager_fieldbook_web');
+
+			insert into permission(name, description, parent_id, workbench_sidebar_category_link_id)
```
Comment (darla-leafnode) [inline src/main/resources/liquibase/workbench_changelog/15_4_0.xml:36]: Can't we just have null values for `workbench_sidebar_category_link_id`?

File: src/main/java/org/generationcp/middleware/dao/GermplasmStudySourceDAO.java (hunk 1 above)
Comment (abatac) [reply src/main/java/org/generationcp/middleware/dao/GermplasmStudySourceDAO.java:61]: I created a new changeset to add unique index constraint for germplasm_study_source.gid.

File: src/main/java/org/generationcp/middleware/domain/inventory/manager/LotDepositRequestDto.java (hunk 2 above)
Comment (abatac) [reply src/main/java/org/generationcp/middleware/domain/inventory/manager/LotDepositRequestDto.java]: Done

File: src/main/java/org/generationcp/middleware/dao/GermplasmStudySourceDAO.java (hunk 3 above)
Comment (abatac) [reply src/main/java/org/generationcp/middleware/dao/GermplasmStudySourceDAO.java]: Done

File: src/main/java/org/generationcp/middleware/service/impl/inventory/TransactionServiceImpl.java (hunk 4 above)
Comment (abatac) [reply src/main/java/org/generationcp/middleware/service/impl/inventory/TransactionServiceImpl.java]: Done

File: src/main/resources/liquibase/workbench_changelog/15_4_0.xml (hunk 5 above)
Comment (abatac) [reply src/main/resources/liquibase/workbench_changelog/15_4_0.xml:36]: I populated this for consistency with other permission records.

**Pull request merged**